### ISSUE TITLE
singularity pull doesn't respect environment variable $SINGULARITY_CACHEDIR 

### ISSUE THREAD

**Issue by satra** [apptainer/singularity#464]
singularity src hash: #3a2155ed

note that the output is placed in `$HOME` and not `$SINGULARITY_CACHEDIR`

```bash
$ echo $SINGULARITY_CACHEDIR 
/om/user/satra/

$ singularity pull shub://singularityhub/ubuntu:14.04
Found image singularityhub/ubuntu:14.04
Downloading image... 5a1d85b1268750b862eea4de157f035f37f21e38.img.gz
Decompressing /home/satra/5a1d85b1268750b862eea4de157f035f37f21e38.img.gz
```

**Comment by satra** [comment]
however, singularity run does the correct thing:

```bash
$ singularity run shub://singularityhub/ubuntu:14.04
Cache folder set to /om/user/satra/shub
Found image singularityhub/ubuntu:14.04
Downloading image... 5a1d85b1268750b862eea4de157f035f37f21e38.img.gz
Decompressing /om/user/satra/shub/5a1d85b1268750b862eea4de157f035f37f21e38.img.gz
No Singularity runscript found, executing /bin/sh
Singularity.5a1d85b1268750b862eea4de157f035f37f21e38.img> $
```

**Comment by vsoch** [comment]
Thanks @satra ! Yes, I originally implemented the `pull` to go into cache, but @gmkurtzer pointed out that it was more intuitive to "pull" to the user's present working directory. The call works via a variable, which could easily be set by the singularity command line tool based on user input/preference. I think what is needed is simply to add ability for the user to specify this. @gmkurtzer what are your thoughts? 

- Should we have another environment variable that is specific to pull, or honor the user's cache directory (as originally implemented)
- Should we have a default for pull, and then a (command line) override?
- Some combination of the two?

As a user, a pull is intuitive to me to (by default) come into my $PWD, but as @satra pointed out, other users might expect it to come to wherever they (or some admin) has defined as the cache, and then return the path. I would also see something reasonable being like:

      singularity pull shub://vsoch/singularity-images newimage.img

Meaning that maybe by default the pull goes to cache, unless the user specifies elsewhere.

**Comment by satra** [comment]
@vsoch - i went by the help string for pull, which may not have been updated :)

in keeping with `docker pull`, which goes into a magical place because of layers, it may be ok to have everything default around some shub directory, whether defaults to `$HOME/shub` or to `$SINGULARITY_ROOT/shub` (note my new env suggestion - all depends on what else may end up going there)

**Comment by vsoch** [comment]
yep, that was definitely my #oups :)

The cache for singularity hub is (by default) set to `$HOME/.singularity/shub` unless the admin/user changes the environment variable. We will definitely think about this, and come up with a more robust and customizable solution for the next release, because as a user I'd want to be able to specify wanting "magic" or just a thing to dump where I'm sitting on the command line! 

Thanks for finding this, and it gives me immense happiness that you found the singularity hub endpoint and gave it a go! :smiley: If you have any feedback about singularity hub, please send my way!

**Comment by satra** [comment]
@vsoch - we are trying to move our cluster users to use `singularity` this year and i will likely do a tutorial sometime next month. if there are teaching materials, for general scientists and neuroscientists, please let me know. i've already interacted with @gmkurtzer to use a vagrant box inside our cluster to create singularity containers, but being able to push to hub and/or pull from the hub will be really helpful to many users.

my next step is try a tensorflow GPU container - that would satisfy many of our users!

**Comment by vsoch** [comment]
Hey @satra ! It would be very reasonable to install singularity on your cluster (or have users install it locally, or use within a docker container) and then have them do a simple tutorial to work with containers. Right now, the most that I have for docs (specific to Singularity Hub) is on the [usage](https://singularity-hub.org/faq) page, and I am giving a talk about Singularity Hub next week and could send along the slides (that have a nice diagram of the workflow).

I really like the idea of some kind of "getting started" tutorial for Singularity (and Singularity Hub). The way that one works with Singularity Hub is by connecting a build specification file (Singularity, akin to Dockerfile) in a Github repo to the hub. Then it builds for them and that's that. So I would imagine something like this:

1. User has local machine with Singularity (within a VM if needed)
2. The user pulls an image, and learns how to interact with it
3. The user then learns that he/she can make this exact image, and is given a base template to do this.
4. The user tries building the image locally, learns about the runscript, post commands, etc (you can give small tasks to do like "install python, print hello world, etc."
5. Finally (if wanted) the user pushes the spec file to a repo, it's built on Singularity Hub, and then he/she can log into your cluster environment and try running the same image.

What are your thoughts on (online) interactive lessons with Asciinema? If you are giving a talk / presentation sort of thing, it might be easier to give them a URL with a step by step tutorial (with ascineema and what not) and then they can move through it at their own pace.

I can definitely help out with this if you would like to work together! It's a bit busy over here, but this doesn't seem like a huge under-taking, and it would be useful for many at Stanford and other places too :)

Everyone wants tensorflow over here too!! Seems like the learning is going deep from East to West :)

**Comment by satra** [comment]
@vsoch - perhaps we can hack this on a google doc - which can then be turned into both info on the singularity site and into an asciinema playback file.

for the purpose of the tutorial at mit it needs to have some specific pieces for our cluster, which allows running a vagrant VM, but requires a hack to make a singularity image (i need to test if this is still true #206). also we need to demonstrate how singularity containers can use GPUs.

so i like the list, but perhaps we can work on the different sections in a google doc to all for a bit more fluid editing.

**Comment by satra** [comment]
just an fyi that the issue from #206 still exists on our cluster.

**Comment by vsoch** [comment]
ok! I need to restart my computer to test it out (this is a rare occurrence, heh) but I should be able to do that soon. If you are intending to install singularity on your cluster, I would just skip the vagrant part and have your users run Singularity.

I got started on some asciinema and (really basic) docs [if you want to look](http://singularity.lbl.gov/singularity-tutorial) and then contribute via [PR to the repo](https://github.com/singularityware/singularityware.github.io/blob/master/_posts/recipes/2017-01-26-singularity-hub-tutorial.md)! We can also do Google Doc, but there isn't nice code formatting and stuffs there :P

**Comment by satra** [comment]
@vsoch - can't give users sudo privileges on cluster - they can run singularity, but i don't think they can create images without sudo, so unless things have changed they have to use a VM to create the image on the cluster.

repo looks good - i'll send a PR to that.

my next step is to figure out how to get the contrib tensorflow + gpu example working on our cluster via a VM. i.e create the image via VM, then run outside on cluster on GPU. currently some of the steps won't work because there is no gpu inside the vm itself.

**Comment by truatpasteurdotfr** [comment]
@satra you don't need a gpu to build a tf gpu enabled image, afaik, you only need to have the matching nvidia drivers version in the containers and on your cluster.

**Comment by satra** [comment]
@truatpasteurdotfr - indeed. in the example, the post-test won't work though. 

i'm just taking a step at a time. however, even python is not installing correctly right now in this example: https://github.com/singularityware/singularity/blob/master/examples/contrib/ubuntu16-tensorflow-0.12.1-gpu.def

**Comment by vsoch** [comment]
+1, I've been working on this! I've already moved through the issues that you noted above - there was definitely some weirdness with the old example version (that I think used to work). I'll ping later with some images to share, I'm going back to sleep for a few hours.

**Comment by satra** [comment]
@vsoch - just sent a PR fixing the issue. https://asciinema.org/a/100998

**Comment by gmkurtzer** [comment]
Hey everyone, what is the status of this?

**Comment by vsoch** [comment]
This should now honor the `SINGULARITY_CACHEDIR`:

```bash
vanessa@vanessa-ThinkPad-T450s:~/Desktop$ export SINGULARITY_CACHEDIR=/tmp
vanessa@vanessa-ThinkPad-T450s:~/Desktop$ singularity --debug pull shub://vsoch/singularity-images
enabling debugging
ending argument loop
Exec'ing: /usr/local/libexec/singularity/cli/pull.exec shub://vsoch/singularity-images
DEBUG:python:SINGULARITY_COMMAND_ASIS found as False
DEBUG:python:SINGULARITY_ROOTFS not defined (None)
DEBUG:python:SINGULARITY_METADATA_FOLDER found as None/.singularity
DEBUG:python:SINGULARITY_DISABLE_CACHE found as False
DEBUG:python:SINGULARITY_CACHEDIR found as /tmp
DEBUG:python:SINGULARITY_ENVBASE found as None/.singularity/env
DEBUG:python:SINGULARITY_LABELFILE not defined (None)
DEBUG:python:SINGULARITY_PULLFOLDER found as /tmp
DEBUG:python:SINGULARITY_CONTENTS found as /tmp/.singularity-layerfile.AfMZs4
INFO:python:
*** STARTING SINGULARITY PYTHON PULL ****
DEBUG:python:SINGULARITY_CONTAINER found as shub://vsoch/singularity-images
DEBUG:python:SINGULARITY_PULLFOLDER found as /tmp
DEBUG:python:Found uri shub://
DEBUG:python:Found uri shub://
INFO:python:Singularity Hub Image: vsoch-singularity-images-mongo.img.gz
INFO:python:Singularity Hub Image: vsoch-singularity-images-mongo.img.gz
Found image vsoch/singularity-images:mongo
Downloading image... vsoch-singularity-images-mongo.img.gz
Decompressing /tmp/vsoch-singularity-images-mongo.img.gz
INFO:python:Singularity Hub Image Download: /tmp/vsoch-singularity-images-mongo.img
DEBUG:python:Writing Singularity Hub image path to /tmp/.singularity-layerfile.AfMZs4
INFO:python:Writing file /tmp/.singularity-layerfile.AfMZs4 with mode w.
Done. Container is at: /tmp/vsoch-singularity-images-mongo.img
```

but note it doesn't work if I export `SINGULARITY_PULLFOLDER`, which should given that I want to pull NOT to PWD, but not to cache.  We need another check here: https://github.com/singularityware/singularity/blob/development/libexec/image-handler.sh#L87 that first checks for the `SINGULARITY_PULLFOLDER`, and then for the cache directory, and only then falls back to PWD.

Does this flow make sense though? I wouldn't want to necessarily pull to cache first before PWD, I would want to have to explicitly define the pull folder or else pull to present working directory. In my mind, the pull folder is a variable that is like a user specified place to put images (and the user could choose his/her cache directory by exporting the same) and the cache directory could be set by the cluster.  So I'm not sure that I completely like the current implementation what about:

 - pull to `SINGULARITY_PULLFOLDER` if defined (note it can be same as cachedir
 - if not defined, go to PWD.

This way, a user working on some cluster that has a shared cache can still have the option to pull to PWD or some custom folder without needing to change the `SINGULARITY_CACHEDIR`. Thoughts?

**Comment by ArangoGutierrez** [comment]
This issue was solved with PR #855 
 > Feel free to reopen
